dotCMS/src/main/java/com/dotmarketing/portlets/languagesmanager/business/LanguageFactoryImpl.java

```diff
@@ -588,13 +590,23 @@ protected void saveLanguageKeys(Language lang, Map<String, String> generalKeys,
 	@Override
 	protected Language getFallbackLanguage(final String languageCode) {
 
+		Language lang = CacheLocator.getLanguageCache().getLanguageByCode(languageCode, "");
+		if (null != lang ) {
+			return (LANG_404.equals(lang)) ? null : lang;
+		}
+
 		try {
 
-			return fromDbMap(new DotConnect()
+			lang = fromDbMap(new DotConnect()
 					.setSQL(SELECT_LANGUAGE_BY_LANG_CODE_ONLY)
 					.addParam(languageCode.toLowerCase())
 					.loadObjectResults().stream().findFirst().orElse(null));
 
+			if(lang == null){
+				CacheLocator.getLanguageCache().add404Language(languageCode, "");
```

Comment:
This is good but I do not see the invalidation when a new language is added?

Reply:
Already happens here: https://github.com/dotCMS/core/blob/4fc93c67146e09f0021c3ad2c47dbce88703e772/dotCMS/src/main/java/com/dotmarketing/portlets/languagesmanager/business/LanguageCacheImpl.java#L145